### PR TITLE
Enable Hunter MM Core for TWW 11.05

### PR DESCRIPTION
**Pull request by Mcheung7272**
### Description

This is a PR to kick off the work needed to get Hunter MM back up to par. This one is soley focused on getting the core aspects of the analyzer available to users (Timeline, Foundation Guide, Death recaps etc). The more nuanced logic in regards to TWW 11.05 patch needs to be refined. There's definitely some stale skill logic still in there and some missing ones from the 11.05 overhaul. 

### Testing

Currently Hunter MM is unavailable as a whole. With this pr, you should be able to analyze Hunter MM logs again.

- Test report URL: 
`/report/ZCjp1DgdVfbxnRBK/26-Mythic+The+Bloodbound+Horror+-+Kill+(3:23)/Pbpbman/standard/debug` - Dark Ranger Test
`/report/AqYvnpjk2MCHmgca/21-Mythic+The+Bloodbound+Horror+-+Kill+(5:10)/Yellot/standard/debug` - Sentinel Test
- Screenshot(s):

image

image

### CODE REVIEW

File: src/analysis/retail/hunter/marksmanship/CHANGELOG.tsx [hunk 1]
```diff
@@ -1,18 +1,22 @@
...
   change(date(2023, 5, 9), <>Added support for T29 tier sets</>, Swolorno),
   change(date(2023, 1, 30), 'Fixed a crashing bug in the Checklist due to a no-longer generated statistic.', emallson),
   change(date(2023, 1, 30), <>Updated <SpellLink spell={TALENTS.CALLING_THE_SHOTS_TALENT} /> to be 2.5 seconds per 50 focus as of 10.0.5.</>, Putro),
   change(date(2023, 1, 30), <>Added <SpellLink spell={TALENTS.DEATHBLOW_TALENT} /> tracking to <SpellLink spell={TALENTS.KILL_SHOT_SHARED_TALENT} /> module.</>, Putro),
-  change(date(2023, 1, 30), <>Fixed an issue with <SpellLink spell={TALENTS.SERPENT_STING_TALENT}/> when combined with <SpellLink spell={TALENTS.SERPENTSTALKERS_TRICKERY_TALENT} /> resulting in incorrect warnings about refreshing.</>, Putro),
+  change(date(2023, 1, 30), <>Fixed an issue with Serpent Sting Talent when combined with <SpellLink spell={TALENTS.SERPENTSTALKERS_TRICKERY_TALENT} /> resulting in incorrect warnings about refreshing.</>, Putro),
   change(date(2022, 12, 23), 'Enable Marksmanship for Dragonflight analysis', Putro),
   change(date(2022, 11, 11), 'Initial transition of Marksmanship to Dragonflight', [Arlie, Putro]),
```
Comment (emallson) [inline src/analysis/retail/hunter/marksmanship/CHANGELOG.tsx]: its okay to remove these old changelog entries (encouraged, even). doesn't need to happen to merge this, but an fyi for the future.

File: src/interface/SpellIcon.tsx [hunk 2]
```diff
@@ -29,7 +29,7 @@ const SpellIcon = ({ spell, noLink, alt, ilvl, ...others }: Props) => {
     />
   );
 
-  const spellId = typeof spell === 'number' ? spell : spell.id;
+  const spellId = typeof spell === 'number' ? spell : spell?.id;
```
Comment (emallson) [inline src/interface/SpellIcon.tsx]: not sure what was being worked around by this, but this is not the correct fix. as per the type: `spell` is a `number` or `Spell`. it should never be null/undefined---although that can happen due to some JS code. the correct fix is to fix the component that is passing `null`/`undefined` here.

im assuming this was just some old `Analyzer` passing in bad data from a spell that no longer exists and probably won't trigger again if it this change is reverted, but ill double check.

EDIT: don't see any broken places in MM now with this removed. if there are specific ones, I can take a look at fixing the callsites.

**Review by emallson — COMMENTED** [review]
changes overall look good. going to throw in a revert of the `SpellIcon` change.

I see some remaining cooldown issues with Rapid Fire and Aimed Shot in the sample log. Are those resets / cooldown shenanigans related to Trueshot that are not yet handled? or other potentially-undetectable resets?

File: src/interface/SpellIcon.tsx (hunk 2 above)
Comment (Mcheung7272) [reply src/interface/SpellIcon.tsx]: I'm not encountering the issue either now. Maybe it was a weird compilation problem. All good with this revert.

File: src/analysis/retail/hunter/marksmanship/CHANGELOG.tsx (hunk 1 above)
Comment (Mcheung7272) [reply src/analysis/retail/hunter/marksmanship/CHANGELOG.tsx]: Good to know. Didnt know if it needed to be kept intact in a sort of audit fashion. Removed those two changelog lines.

Comment (emallson) [reply src/analysis/retail/hunter/marksmanship/CHANGELOG.tsx]: i usually nuke the entire changelog at the start of an expansion for the specs that i work on :P i didn't for a while but 3 expansions in the list gets pretty damn long

**Comment by Mcheung7272** [comment]
> changes overall look good. going to throw in a revert of the `SpellIcon` change.
> 
> I see some remaining cooldown issues with Rapid Fire and Aimed Shot in the sample log. Are those resets / cooldown shenanigans related to Trueshot that are not yet handled? or other potentially-undetectable resets?

Not entirely sure tbh. I have a feeling it might have to do with [Surging Shots](https://www.wowhead.com/spell=336867/surging-shots) maybe? I don't see us tracking the possibility of rapidfire resetting by 15%. Theres still a bunch of TWW logic to be added tbh

**Comment by emallson** [comment]
yup, that looks like an undetectable reset all right.

if there is a buff that is granted when you get a surging shots reset, you can use that. otherwise, this is usually handled by extending the core `SpellUsable` class, checking if it is on cooldown when you cast it, and assuming there is was proc at the last possible point (so, most recent Aimed Shot cast?) that reset the cooldown. i don't know of a spec offhand that has that implemented right now, but you could look at subclasses of `SpellUsable` and find one.

we can merge without that. that seems like a fairly minor issue as long as there is a reason known for it being off

**Comment by Mcheung7272** [comment]
I'm all good with a merge if you are. Thanks for the pointer, ill take a look. I'm hoping as I start adding more missing logic, these cooldown issues will naturally iron out.

**Pull request merged**